server/src/services/library.service.ts

```diff
@@ -598,106 +622,96 @@ export class LibraryService extends EventEmitter {
+      existingAssetsDone = existingAssetPage.done ?? true;
+
+      if (existingAssetPage.value) {
+        existingAssetCounter += existingAssetPage.value.length;
+        this.logger.log(
+          `Queuing online check of ${existingAssetPage.value.length} asset(s) in library ${library.id}...`,
+        );
+        await this.jobRepository.queueAll(
+          existingAssetPage.value.map((asset: AssetEntity) => ({
+            name: JobName.LIBRARY_CHECK_OFFLINE,
+            data: { id: asset.id, importPaths: validImportPaths },
+          })),
+        );
       }
-    }
+    };
 
-    this.logger.verbose(`Crawled assets have been checked for online/offline status`);
+    let crawledAssetPaths: string[] = [];
 
-    if (assetIdsToMarkOffline.length > 0) {
-      this.logger.debug(`Found ${assetIdsToMarkOffline.length} offline asset(s) previously marked as online`);
-      await this.assetRepository.updateAll(assetIdsToMarkOffline, { isOffline: true });
-    }
+    while (!crawlDone) {
+      const crawlResult = await crawledAssets.next();
 
-    if (assetIdsToMarkOnline.length > 0) {
-      this.logger.debug(`Found ${assetIdsToMarkOnline.length} online asset(s) previously marked as offline`);
-      await this.assetRepository.updateAll(assetIdsToMarkOnline, { isOffline: false });
-    }
+      crawlDone = crawlResult.done ?? true;
 
-    if (crawledAssetPaths.size > 0) {
-      if (!shouldScanAll) {
-        this.logger.debug(`Will import ${crawledAssetPaths.size} new asset(s)`);
+      if (!crawlDone) {
+        crawledAssetPaths.push(crawlResult.value);
+        crawlCounter++;
       }
 
-      let batch = [];
-      for (const assetPath of crawledAssetPaths) {
-        batch.push(assetPath);
+      if (crawledAssetPaths.length % LIBRARY_SCAN_BATCH_SIZE === 0 || crawlDone) {
+        this.logger.log(`Queueing scan of ${crawledAssetPaths.length} asset path(s) in library ${library.id}...`);
+        // We have reached the batch size or the end of the generator, scan the assets
+        await this.scanAssets(job.id, crawledAssetPaths, library.ownerId, job.refreshAllFiles ?? false);
+        crawledAssetPaths = [];
 
-        if (batch.length >= LIBRARY_SCAN_BATCH_SIZE) {
-          await this.scanAssets(job.id, batch, library.ownerId, job.refreshAllFiles ?? false);
-          batch = [];
+        if (!existingAssetsDone) {
+          // Interweave the queuing of offline checks with the asset scanning (if any)
+          await checkIfOnlineAssetsAreOffline();
         }
       }
-
-      if (batch.length > 0) {
-        await this.scanAssets(job.id, batch, library.ownerId, job.refreshAllFiles ?? false);
-      }
     }
 
-    await this.repository.update({ id: job.id, refreshedAt: new Date() });
-
-    return JobStatus.SUCCESS;
-  }
+    // If there are any remaining assets to check for offline status, do so
+    while (!existingAssetsDone) {
+      await checkIfOnlineAssetsAreOffline();
+    }
```

Comment:
This can easily become an infinite loop. Let's use a for loop instead.

Reply:
See previous comment